File: LeetCode/Arrays/rotate-array.c
/*
 * I've been stuck on this problem for the past couple of times I tried.
 * I was contistently having an issue with moving the elements with a single
 * swap element to meet the O(1) by moving the elements based on k progress.
 * However I had to read the solution to remember that O(1) memory is just
 * fixed regarless of the size of N, as long is not O(N) or O(k).
 *
 * https://leetcode.com/problems/rotate-array/
 * 
 * Runtime: 141 ms (62.91% Better)
 * Memory Usage: 22.2 MB (58.87 Better than)
 * 
 *  https://leetcode.com/PlaidCat/
 */

void print_array(int*nums, int numsSize){
  int i = 0;
  for (i ; i < numsSize; ++i){
    printf("%i,", nums[i]);
  }
  printf("\n");
}


void rotate(int* nums, int numsSize, int k){
  int i = 0;
  int j = 0;
  int p = 0;
  int count = 0;
  int swap = 0;
  k %= numsSize;

  
  for (i = 0; count < numsSize; ++i){
    int cur = i;

    swap = nums[cur];
    do {
      int next = (cur + k) % numsSize;
      int tmp = nums[next];
      //printf("i: %i =?= ", i);
      //printf("cur: %i\n", cur);
      //printf("next: %i\n", next);
      //printf("swap: %i\n", swap);
      //printf("tmp: %i\n", tmp);
      nums[next] = swap;
      swap = tmp;
      cur = next;
      //printf("cur: %i\n", cur);
      count++;
    }while (cur != i);
    //print_array(nums, numsSize);
  }
}

File: LeetCode/Arrays/rotate-array.rs
/*
 * Based this off the basic C program in the same dir.
 * 
 * A lot of this was trying to take something that is very easy for me to understand
 * and put it into the voodoo of the rust compiler.
 *
 * https://leetcode.com/problems/rotate-array/
 * 
 * Runtime: 20 ms (29.68% Better)
 * Memory Usage: 3.6 MB (87.70 Better than)
 * 
 *  https://leetcode.com/PlaidCat/
 */

impl Solution {
  pub fn rotate(nums: &mut Vec<i32>, k: i32) {
    let len = (nums.len());
    let my_k = ((k as usize) % len);
    let mut count = 0;
    let mut i: usize = 0;
    
    while count < nums.len() {
      let mut cur: usize = i;
      //println!("nums[{}] = {}", cur, nums[cur]);
      let mut swap = nums[cur];
      loop {
        let next: usize = (cur + my_k) % len;
        let tmp = nums[next];
        //println!("Cur: {} -> Next: {} == k: {}\n", cur, next, my_k);
        nums[next] = swap;
        swap = tmp;
        cur = next;
        count += 1;
        if cur == i { break; }
      } 
      i+=1;
    }
  }
}